lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,879 @@
+        Ok(())
+    }
+
+    fn write_entry<W: Write>(
+        &self,
+        mut writer: W,
+        external_id: &PointIdType,
+    ) -> OperationResult<()> {
+        // Serializing External ID
+        match external_id {
+            PointIdType::NumId(num) => {
+                // Byte to distinguish between Number and UUID
+                writer.write_u8(Self::EXTERNAL_ID_NUMBER_BYTE)?;
+
+                // The PointID's number
+                writer.write_u64::<FileEndianess>(*num)?;
+            }
+            PointIdType::Uuid(uuid) => {
+                // Byte to distinguish between Number and UUID
+                writer.write_u8(Self::EXTERNAL_ID_UUID_BYTE)?;
+
+                // The PointID's UUID
+                writer.write_u128::<FileEndianess>(uuid.to_u128_le())?;
+            }
+        }
+
+        let internal_id = match external_id {
+            PointIdType::NumId(n) => self.external_to_internal_num.get(n),
+            PointIdType::Uuid(u) => self.external_to_internal_uuid.get(u),
+        }
+        .ok_or(OperationError::PointIdError {
+            missed_point_id: *external_id,
+        })?;
+
+        // Serializing Internal ID
+        writer.write_u32::<FileEndianess>(*internal_id)?;
+
+        Ok(())
+    }
+}
+
+impl ImmutableIdTracker {
+    pub fn open(segment_path: &Path) -> OperationResult<Self> {
+        let deleted_raw = open_write_mmap(&Self::deleted_file_path(segment_path))?;
+        let deleted_mmap = MmapBitSlice::try_from(deleted_raw, 0)?;
+        let deleted_bitvec = deleted_mmap.to_bitvec();
+        let deleted_wrapper = MmapBitSliceBufferedUpdateWrapper::new(deleted_mmap);
+
+        let internal_to_version_map =
+            open_write_mmap(&Self::version_mapping_file_path(segment_path))?;
+        let internal_to_version_mapslice: MmapSlice<SeqNumberType> =
+            unsafe { MmapSlice::try_from(internal_to_version_map)? };
+        let internal_to_version = internal_to_version_mapslice.to_vec();
+        let internal_to_version_wrapper =
+            MmapSliceBufferedUpdateWrapper::new(internal_to_version_mapslice);
+
+        let reader = BufReader::new(File::open(Self::mappings_file_path(segment_path))?);
+        let mappings = PointMappings::load(reader, Some(&deleted_bitvec))?;
+
+        Ok(Self {
+            path: segment_path.to_path_buf(),
+            deleted: deleted_bitvec,
+            deleted_wrapper,
+            internal_to_version_wrapper,
+            internal_to_version,
+            mappings,
+        })
+    }
+
+    pub(super) fn new(
+        path: &Path,
+        deleted: &BitSlice,
+        internal_to_version: &[SeqNumberType],
+        mappings: PointMappings,
+    ) -> OperationResult<Self> {
+        // Create mmap file for deleted bitvec
+        let deleted_filepath = Self::deleted_file_path(path);
+        {
+            let deleted_size = bitmap_mmap_size(deleted);
+            create_and_ensure_length(&deleted_filepath, deleted_size)?;
+        }
```

Comment:
`create_and_ensure_length` returns a file-handle. To prevent issues when opening the memory-map again in the next line, I thought flushing/dropping it first would be safer.
It is also scoped here: https://github.com/qdrant/qdrant/blob/master/lib/segment/src/vector_storage/chunked_mmap_vectors.rs#L55
But maybe it is not really needed?

Reply:
It'll be dropped right away anyway because we don't store the resulting value in a variable.

But feel free to leave it like this to _emphasize_ this a bit.